### ISSUE TITLE
Documentation: Need example on how to dynamically link an application

### ISSUE THREAD

**Issue by jrmiller82** [therecipe/qt#401]
Since QT is lgpl, this library is pretty much unusable without being able to dynamically link QT to the go application. Your readme.md implies that dynamic linking is available for certain platforms. An example way to build such a dynamically linked executable would be helpful since golang is typically known for statically linked binaries.

**Comment by jrmiller82** [comment]
Eeek. Just saw that this app is licensed under the LGPL and not a permissive license like MIT or Apache... how in the heck is a golang user supposed to use this with a commercial distributed application? I smell a PyQT paid license coming...

**Comment by therecipe** [comment]
Hey

qtdeploy always dynamically links against Qt by default unless you explicitly export `QT_MXE_STATIC` or `QT_MSYS2_STATIC`.
The only exception to this is iOS, where statically linking is the only option at the moment.
But this may change with 5.10.

>how in the heck is a golang user supposed to use this with a commercial distributed application?

If the application is closed source (and you dynamically link against Qt) then you would need to either:
* Distribute object files + instructions on how to re-link against them.
* Make your source accessible to your end users (not necessary under an open source license), but so that the end user is able to re-compile the application against another version of this binding (and Qt).
* Not yet official possible: Buy a commercial license (similar to PyQt's)

(I can understand if none of these options is applicable for you, but these are the options)

However if your application is open source, then there are no further conditions that have to be meet.

**Comment by jrmiller82** [comment]
Distributing object files + instructions on how to re-link is most certainly importing the viral qualities of LGPL and I don't wish to be the test lawsuit to determine that.

**Comment by therecipe** [comment]
>Distributing object files + instructions on how to re-link is most certainly importing the viral qualities of LGPL and I don't wish to be the test lawsuit to determine that.

I was under the assumption that "work that uses the library" is not triggering the viral part of LGPL.
Therefor it makes no difference how you comply with these conditions.
See here: https://www.gnu.org/licenses/gpl-faq.en.html#LGPLStaticVsDynamic

**Comment by jrmiller82** [comment]
That's a FAQ and not the license itself. The license itself is not black and white when it comes to that issue.

**Comment by joeblew99** [comment]
Would be good to know what your going to expect to be paid for a license :)

**Comment by joeblew99** [comment]
Pyqt is 600 per developer ?

https://www.riverbankcomputing.com/commercial/buy

Just select t an option and it tells you the cost

**Comment by therecipe** [comment]
@jrmiller82
Here is an excerpt from: https://www.gnu.org/licenses/lgpl-3.0

>d) Do one of the following:

>0) Convey the Minimal Corresponding Source under the terms of this License, and the Corresponding Application Code in a form suitable for, and under terms that permit, the user to recombine or relink the Application with a modified version of the Linked Version to produce a modified Combined Work, in the manner specified by section 6 of the GNU GPL for conveying Corresponding Source.
>1) Use a suitable shared library mechanism for linking with the Library. A suitable mechanism is one that (a) uses at run time a copy of the Library already present on the user's computer system, and (b) will operate properly with a modified version of the Library that is interface-compatible with the Linked Version.

@joeblew99
It will probably be less, but therefore recurring.
And probably also divided into different tiers.

**Comment by joeblew99** [comment]
Ok well that's something at least....

Don't get me wrong. It's good your paid and this is valuable. It's more about risk and knowing that all the code invested on top of QT golang is not going away actually too.

**Comment by RadhiFadlillah** [comment]
Hi @therecipe, first of all, thanks for this package. Just like @jrmiller82, I'm also wondering about restriction from using `Qt` and `therecipe/qt` (which both using `LGPLv3`) on my app. After spending hours reading about LGPLv3 (which after this point will be referred simply as LGPL), I found out about some things, but become more lost on other things.

One of the most asked question about LGPL is: **"Can I use library with LGPL license for my commercial, closed source app ?"**. More specifically, in my case :

- I want to write a closed and commercial app using `Go` language.
- I want to use `Qt` and `therecipe/qt` which uses LGPL, but I **won't** modify those library, and I will only use its API, functions and structs.

To simplify my points, I will separate my research result into two parts.

First I will talk about `Qt` library.

LGPL license [insists](https://www.gnu.org/licenses/lgpl.html) that end user must be able to recombine or relink the application with a modified or different version of the LGPL library, to produce a modified executable :

```
0) Convey the Minimal Corresponding Source under the terms of this License, and the 
Corresponding Application Code in a form suitable for, and under terms that permit, the 
user to recombine or relink the Application with a modified version of the Linked Version 
to produce a modified Combined Work, in the manner specified by section 6 of the GNU 
GPL for conveying Corresponding Source.
```

This can be achieved using [two method](http://answers.google.com/answers/threadview/id/439136.html) :

- By **static linking** the library to my executable, but still giving its source code or object code so user can recompile my app using different version of library.
- By **dynamic linking** the library to my executable, so user can easily change or install which version of library that he wants.

Because of my use case, and the fact that (AFAIK) `Go` doesn't make an object code, so my only choice is by using **dynamic linking**. Fortunately `qtdeploy` by default already dynamic links to `Qt` library, so I can just distribute the executable and user can install his own library.

So, from what I found `Qt` is passable for my use case. Now the problem is on `therecipe/qt`.

Package `therecipe/qt` composed by two parts :

- [Tools](https://github.com/therecipe/qt/tree/master/cmd) like `qtsetup`, `qtdeploy`, `qtrcc`, `qtmoc` and `qtminimal` which used for (CMIIW) generating code for inherited class and compiling application into executable.
- API that binding `Go` to `Qt` class, like `widgets`, `quick`, `core`, etc.

Both tools and API is released under LGPL license. Now, what is the implication of using these tools and API for creating my closed and commercial app ?

[GPL FAQ](https://www.gnu.org/licenses/gpl-faq.html) explains that apps or tools with GPL and LGPL license [can be used](https://www.gnu.org/licenses/gpl-faq.html#CanIUseGPLToolsForNF) for developing and compiling closed source app :

```
Can I use GPL-covered editors such as GNU Emacs to develop nonfree programs? 
Can I use GPL-covered tools such as GCC to compile them?

Yes, because the copyright on the editors and tools does not cover the code you write. 
Using them does not place any restrictions, legally, on the license you use for your code.
```

Now, the only problem left is with `therecipe/qt` API.

Repeating above, LGPL license insists that end user must be able to recombine or relink the application with a modified or different version of the LGPL library, to produce a modified executable.

As we know, `Go` compile its code into one single executable file. Now, what happen if end user want to change the version of `therecipe/qt`. For example, I deploy my app on 21 June 2017 using commit `5b0427c`, but when my user use my app on 20 July 2017, the latest version of `therecipe/qt` is `4e40d6b`, so my user want to change his app, **on his own**, to use the latest version of `therecipe/qt`.

According to LGPL license, I have to comply with his wish. And since `Go` by default static linking all imported package, and `Go` (AFAIK) doesn't have object code, my only choice is giving him my source code, which is not allowed on my case.

The problem with `Go` and LGPL also discussed [here](https://groups.google.com/forum/#!topic/golang-nuts/JqOAWBpL-70).

Now that I finished explaining about my finding and problems, the only thing left is questions :

1. Is my explanation above correct, or maybe there are some parts that I misunderstand about these LGPL stuff ? Is `therecipe/qt` really not suitable for developing closed commercial app ?
2. I'm sorry for asking, but is there any reason that make you use LGPL license for this package ? I mean, by design `Go` will alway static linking all libraries that used in an app into a single executable, and it looks like it will not change anytime soon.
3. Is there any possibility for changing the license into something more permissive like MIT, BSD, etc ?
4. While I'm on it, do you have plan to create a new name for this package ? It kinda weird to say `qt is Go binding for Qt`, and `therecipe/qt` is too long to write :)

Alright, that's all, thanks for your answer.

**Comment by jrmiller82** [comment]
@RadhiFadlillah : You'll have to buy @therecipe  forthcoming commercial license subscription as he/she has stated if you want to keep your source closed.

**Comment by therecipe** [comment]
@RadhiFadlillah 

>Because of my use case, and the fact that (AFAIK) Go doesn't make an object code

There is object code created, but it's "hidden" inside some temp folder.
You can preserve it by running: `go build -work -x` (then look into the "WORK" folder)

The last 10-15 lines of the output are also what your user would need to run to re-link your object code against another version of the binding.

>Both tools and API is released under LGPL license. Now, what is the implication of using these tools and API for creating my closed and commercial app ?

Afaik, code created with qtmoc and qtrcc is yours.
Because the code created thereby is (probably) heavily derived from your instructions.

But code created by qtsetup and qtminimal is still under LGPL.
Because the code created is already explicitly declared as LGPL in this repo.

And qtdeploy uses qtrcc, qtmoc and qtminimal under the hood, so it's a mix.

>Is my explanation above correct, or maybe there are some parts that I misunderstand about these LGPL stuff ? Is therecipe/qt really not suitable for developing closed commercial app ?

Yeah, your explanations are correct afaik.
And yes, this binding is not (easily) suitable to be used in closed source application that are intended to be distributed to the general public (officially) atm. (They don't even have to be commercial nature)

That's why I opened https://github.com/therecipe/qt/issues/259, to be able to add another license that allows the easy use of this project in a closed source application.

>I'm sorry for asking, but is there any reason that make you use LGPL license for this package ? I mean, by design Go will alway static linking all libraries that used in an app into a single executable, and it looks like it will not change anytime soon.

Yeah, currently LGPL is still needed to secure the further development of this project.
I hoped that the Go authors would get dynamic linking against Go code working in Go >= 1.6, so that the LGPL license wouldn't be to problematic. But sadly they only made it work on linux yet.

>Is there any possibility for changing the license into something more permissive like MIT, BSD, etc ?

Not atm, but maybe someday in the future.

>While I'm on it, do you have plan to create a new name for this package ? It kinda weird to say qt is Go binding for Qt, and therecipe/qt is too long to write :)

Yeah, it's planned.
But I have no real idea for a name yet.
See here: https://github.com/therecipe/qt/issues/382


To sum it up, there will be a commercial license subscription in the future to make this project suitable to be used in closed source (+commercial) applications, like @jrmiller82 said.

**Comment by RadhiFadlillah** [comment]
Hey @therecipe thanks for your answer. 

First of all, I support your plan to make commercial license for this binding and I believe it will be successful. `Go` is very easy to learn and [suitable](https://texlution.com/post/why-go-is-doomed-to-succeed/) for usage within organization thanks to its opinionated design. On the other hand, `Qt` currently has become the standard GUI framework for native app. Using `therecipe/qt` with its commercial license, companies can easily build native GUI app, that can be cross compiled to almost all platform, and can be run from only one single executable, which in turn will make distribution to end user very easy.

Since you are serious about making commercial license, I suggest you to change the name of this binding as soon as possible, especially since `Qt` is already trademarked.

Once again, thanks for your answer and good luck to you @therecipe.

**Comment by therecipe** [comment]
@RadhiFadlillah
Thanks a ton for your work on the wiki and also for your support :) I really appreciate it.

I initially didn't planned to make use of a wiki at all, because I never saw one that I really liked.
But this looks already better than the monstrosity of a readme that we currently have.

I'm a busy today, so I will look into it in more detail tomorrow and also help out a bit.
But it looks already awesome from the quick glance I took.

>Since you are serious about making commercial license, I suggest you to change the name of this binding as soon as possible, especially since Qt is already trademarked.

Yeah, I will try to come up with a real name asap.
But I think it's safe to have the repo called "qt" atm, since I also don't promote it as "qt" or so.

**Comment by jrmiller82** [comment]
https://www.qt.io/trademark/
